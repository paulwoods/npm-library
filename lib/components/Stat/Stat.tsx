// noinspection JSUnusedGlobalSymbols

import React from "react";
import {useThemeProps} from "@mui/material";
import {StatProps} from "../../types.tsx";
import {StatRoot} from "../../main.ts";
import {StatValue} from "../../main.ts";
import {StatUnit} from "../../main.ts";

export const Stat = React.forwardRef<HTMLDivElement, StatProps>(function Stat(inProps, ref) {
    const props = useThemeProps({props: inProps, name: 'MuiStat'})
    const {value, unit, variant, ...other} = props;

    const ownerState = {...props, variant}

    return (
        <StatRoot ref={ref} ownerState={ownerState} {...other}>
            <StatValue ownerState={ownerState}>{value}</StatValue>
            <StatUnit ownerState={ownerState}>{unit}</StatUnit>
        </StatRoot>
    )
})
