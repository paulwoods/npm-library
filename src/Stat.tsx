// noinspection JSUnusedGlobalSymbols

import {ComponentsOverrides, ComponentsVariants, styled, Theme as MuiTheme} from "@mui/material/styles";
import React from "react";
import {useThemeProps} from "@mui/material";

interface StatProps {
    value: string
    unit: string
    variant?: string
}

interface StatOwnerState extends StatProps {
}

type Theme = Omit<MuiTheme, 'components'>;

declare module '@mui/material/styles' {
    interface ComponentNameToClassKey {
        MuiStat: 'root' | 'value' | 'unit';
    }

    interface ComponentsPropsList {
        MuiStat: Partial<StatProps>;
    }

    interface Components {
        MuiStat?: {
            defaultProps?: ComponentsPropsList['MuiStat'];
            styleOverrides?: ComponentsOverrides<Theme>['MuiStat'];
            variants?: ComponentsVariants['MuiStat'];
        };
    }
}

const StatRoot = styled('div', {
    name: 'MuiStat',
    slot: 'root'
})<{ ownerState: StatOwnerState }>(({theme, ownerState}) => ({
    display: "flex",
    flexDirection: "column",
    gap: theme.spacing(0.5),
    padding: theme.spacing(3, 4),
    backgroundColor: theme.palette.background.paper,
    borderRadius: theme.shape.borderRadius,
    boxShadow: theme.shadows[2],
    letterSpacing: '-0.025em',
    fontWeight: 600,
    ...ownerState.variant === 'outlined' && {
        border: `2px solid ${theme.palette.divider}`
    }
}))

const StatValue = styled('div', {
    name: 'MuiStat',
    slot: 'value'
})<{ ownerState: StatOwnerState }>((({theme}) => ({
    ...theme.typography.h3
})))

const StatUnit = styled('div', {
    name: 'MuiStat',
    slot: 'unit'
})<{ ownerState: StatOwnerState }>((({theme}) => ({
    ...theme.typography.body2,
    color: theme.palette.text.secondary
})))

export const Stat = React.forwardRef<HTMLDivElement, StatProps>(function Stat(inProps, ref) {
    const props = useThemeProps({props: inProps, name: 'MuiStat'})
    const {value, unit, variant, ...other} = props;

    const ownerState = {...props, variant}

    return (
        <StatRoot ref={ref} ownerState={ownerState} {...other}>
            <StatValue ownerState={ownerState}>{value}</StatValue>
            <StatUnit ownerState={ownerState}>{unit}</StatUnit>
        </StatRoot>
    )
})
