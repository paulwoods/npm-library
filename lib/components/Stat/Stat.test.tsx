import renderer from 'react-test-renderer'
import {describe, expect, it} from "vitest";
import {Stat} from "./Stat.tsx";

describe('Stat component', () => {
    it('renders correctly', () => {
        const component = renderer.create(<Stat unit="units" value="123"/>)
        const tree = component.toJSON()
        expect(tree).toMatchSnapshot()
    })
})