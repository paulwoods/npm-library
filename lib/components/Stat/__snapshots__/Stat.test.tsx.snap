// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Stat component > renders correctly 1`] = `
<div
  className="css-15pprpf-MuiStat-root"
>
  x
  <div
    className="css-jzuw05-MuiStat-value"
  >
    123
  </div>
  <div
    className="css-19z9boo-MuiStat-unit"
  >
    units
  </div>
</div>
`;
